### ISSUE TITLE
x/tools/gopls: friction working with internal/lsp/tests

### ISSUE THREAD

**Issue by findleyr** [golang/go#54845]
This issue catalogs friction experienced with the "old" LSP "marker tests" in internal/lsp/testdata (run by internal/lsp/tests).

These tests have long been tricky to work with, and recently have caused a significant amount of friction while making changes to error messages in go/parser and go/types.

Notable problems:

- [x] test output include noisy LSP logs for the entire test session. It would be better if these logs were scoped to the actual failing package, or completely excluded.
- [x] test output includes red-herring "errors" that are not actually related to the test failure 
- [x] tests are all run in the same session / workspace, so changes in far-away files can affect e.g. completion results
- [x] auto-generated test names (which include the annotation position) are not stable, and confusing
- [x] failure messages can be hard to read, because they do a poor job of highlighting differences between expected and actual output.
- [x] tests often match error messages too precisely, resulting in churn when error messages change across Go versions
- [x] test annotations are not documented; it is not clear how to add new annotations
- [x] tests run in multiple contexts (as tests for the internal/lsp/source, internal/lsp/cmd and gopls packages), and this is not clearly documented, nor are the differences between these contexts made clear. (and the necessity for all three contexts is not clear)
- [x] tests use `summary*.txt.golden` files (depending on go version) as checksums to ensure that the expected number of tests ran. These are (by construction!) change detectors.

CC @adonovan @griesemer

**Comment by findleyr** [comment]
@adonovan and I just discussed where to go with this next, and there are a few major problems we'd like to solve:
1. the global namespace of marker tests: adding a new test can cause action at a distance
2. the lack of discoverability for new markers, and generally poor documentation for marker tests
3. the discrepancy with the regtests (and regtests could be marker tests, but aren't because of 1 and 2)

We just brainstormed a design for solving all of these problems:
- represent the marker, collection method, and application method as an instance of a new type
- run this new type through the regtest framework

To illustrate what I mean, here is a hypothetical implementation of an isolated marker test:

```go
func TestMarkerTest(t *testing.T) {
	const mod = `
-- go.mod --
module mod.com

go 1.12
-- main.go --
package main

var C = 3

func main() {
	println(C) //@rename("C", "D")
}
-- main.go.rename_C_D_unified --
- var C = 3
+ var D = 3
-	println(D) //@rename("C", "D")
+	println(D) //@rename("C", "D")
`
	RunMarkerTest(t, mod, tests.Rename)
	RunMarkerTest(t, mod, tests.Rename, tests.References)
	RunMarkerTest(t, mod, tests.All...)
}
```
